bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/bundlefile/CloseableBundleFile.java

```diff
@@ -141,8 +142,21 @@ private void open(boolean keepLock) throws IOException {
 			} else {
 				mruListUse();
 			}
+		} catch (Error | RuntimeException e) {
+			// if we get here then something unexpected happened (NPE, OOM, etc.);
+			// force the lock to be released
+			forceUnlock = e;
```

Comment:
Can't you just use the existing keepLock and set it to false here? In some ways, if the caller passes keepLock=true, then then caller is really responsible for releasing the lock even if this method terminates exceptionally.

Reply:
Instead of having a boolean passed to `open(keepLock)` I moved the lock responsibilities out to the two callers of this method `open()` and `lockOpen()`.  This way `lockOpen()` will know that it has acquired the lock without any doubt and it can release the lock on all Throwable thrown to the call to the new `internalOpen()` method.